rules/prefer-error-cause.js

```diff
@@ -0,0 +1,364 @@
+	});
+};
+
+const getAllNodesToFix = ({node, context, throwStatement}) => {
+	const result = [];
+
+	let statementToFix;
+	let errorConstructorLastArgument;
+
+	// `try {} catch (err) { throw new Error('oops', {cause: err})}`
+	// `promise.catch(err => { throw new Error('oops', {cause: err})`
+	if (throwStatement.argument.type === 'NewExpression') {
+		errorConstructorLastArgument = peekTop(throwStatement.argument.arguments);
+
+		if (!errorConstructorLastArgument) {
+			reportCannotBeFixed(node, context);
+			return;
+		}
+
+		result.push({
+			node: errorConstructorLastArgument,
+			statementToFix: throwStatement,
+		});
+	} else {
+		// Assume Error's constructor exists in other node of the block
+		// It could be VariableDeclarator or AssignmentExpression.
+
+		const newExpressionSelector = reference =>
+			(reference.identifier?.parent.type === 'AssignmentExpression'
+			|| reference.identifier?.parent.type === 'VariableDeclarator')
+			&& reference.writeExpr?.type === 'NewExpression'
+			&& reference.resolved?.name === throwStatement.argument.name;
+
+		const thrownErrorDeclarators = getReferences(context.getScope())
+			.filter(reference => newExpressionSelector(reference))
+			.map(reference => reference.identifier.parent);
+
+		// Report newExpression not found in given block.
+		// `try {} catch { const err = new Error; try {} catch { throw err; } }`
+		if (_.isEmpty(thrownErrorDeclarators)) {
+			reportCannotBeFixed(node, context);
+			return;
+		}
+
+		// `try {} catch (err) { let e1 = new Error('oops'); e1 = new Error('oops'); throw error;}`
+
+		for (const thrownErrorDeclarator of thrownErrorDeclarators) {
+			statementToFix = thrownErrorDeclarator;
+
+			let targetArguments;
+			if (thrownErrorDeclarator.type === 'VariableDeclarator') {
+				targetArguments = thrownErrorDeclarator.init.arguments;
+			} else if (thrownErrorDeclarator.type === 'AssignmentExpression') {
+				targetArguments = thrownErrorDeclarator.right.arguments;
+			}
+
+			errorConstructorLastArgument = peekTop(targetArguments);
+
+			if (!errorConstructorLastArgument) {
+				reportCannotBeFixed(node, context);
+				return;
+			}
+
+			result.push({
+				node: errorConstructorLastArgument,
+				statementToFix,
+			});
+		}
+	}
+
+	return result;
+};
+
+const handleCatchBlock = ({node, context, statements, parameter}) => {
+	const throwStatement = statements.find(statement => statement.type === 'ThrowStatement');
+
+	// Filter blocks not having throw statement.
+	// `try {} catch (error) {}`
+	// `promise.catch(error => {})`
+	if (!throwStatement) {
+		return;
+	}
+
+	const errorArgumentIdentifier = parameter?.name;
+	const throwStatementArgument = throwStatement.argument;
+
+	// Filter block having valid cause property
+	// `try {} catch (err) { throw new Error('oops', {cause: err})}`
+	// `promise.catch(err => { throw new Error('oops', {cause: err})`
+	if (errorArgumentIdentifier && errorArgumentIdentifier === throwStatementArgument.name) {
```

Comment:
We should not check the name. 


```js
try {} catch (outerError) {
	try {} catch (innerError) {
		throw new Error('', {cause: outerError});
	}
}
```

I think this make sense, just make sure `cause` property exist.